src/libbson/src/bson/bson-atomic.h

```diff
@@ -25,85 +25,285 @@
+                                          int64_t expect_value,
+                                          int64_t new_value,
+                                          enum bson_atomic_memorder);
+
+extern void
+bson_yield_thread ();
+
+#if (defined(_MSC_VER) && !defined(_M_IX86)) || defined(__LP64__)
+/* (64-bit intrinsics are only available in x64) */
+DECL_ATOMIC_STDINT (int64, 64)
 #else
-#define __BSON_NEED_BARRIER 1
-#warning "Unknown compiler, using lock for compiler barrier."
-BSON_EXPORT (void)
-bson_memory_barrier (void);
+static BSON_INLINE int64_t
+bson_atomic_int64_fetch_add (int64_t volatile *val,
+                             int64_t v,
+                             enum bson_atomic_memorder order)
+{
+   return _bson_emul_atomic_int64_fetch_add (val, v, order);
+}
+
+static BSON_INLINE int64_t
+bson_atomic_int64_exchange (int64_t volatile *val,
+                            int64_t v,
+                            enum bson_atomic_memorder order)
+{
+   return _bson_emul_atomic_int64_exchange (val, v, order);
+}
+
+static BSON_INLINE int64_t
+bson_atomic_int64_compare_exchange (int64_t volatile *val,
+                                    int64_t expect_value,
+                                    int64_t new_value,
+                                    enum bson_atomic_memorder order)
+{
+   return _bson_emul_atomic_int64_compare_exchange (
+      val, expect_value, new_value, order);
+}
 #endif
 
+DECL_ATOMIC_INTEGRAL (int, int, )
+
+static BSON_INLINE void *
+bson_atomic_ptr_exchange (void *volatile *ptr,
+                          void *new_value,
+                          enum bson_atomic_memorder ord)
+{
+   DEF_ATOMIC_OP (
+      _InterlockedExchangePointer, __atomic_exchange_n, ord, ptr, new_value);
+}
+
+static BSON_INLINE void *
+bson_atomic_ptr_compare_exchange (void *volatile *ptr,
+                                  void *expect,
+                                  void *new_value,
+                                  enum bson_atomic_memorder ord)
+{
+   switch (ord) {
+   case bson_memorder_release:
+   case bson_memorder_seqcst:
+      DEF_ATOMIC_CMPEXCH (Pointer, , __ATOMIC_SEQ_CST, ptr, expect, new_value);
+      return expect;
+   case bson_memorder_relaxed:
+      DEF_ATOMIC_CMPEXCH (Pointer,
+                          MSVC_MEMORDER_SUFFIX (_nf),
+                          __ATOMIC_RELAXED,
+                          ptr,
+                          expect,
+                          new_value);
+      return expect;
+   case bson_memorder_acquire:
+      DEF_ATOMIC_CMPEXCH (Pointer,
+                          MSVC_MEMORDER_SUFFIX (_acq),
+                          __ATOMIC_ACQUIRE,
+                          ptr,
+                          expect,
+                          new_value);
+      return expect;
+   default:
+      BSON_UNREACHABLE ("Invalid bson_atomic_memorder value");
+   }
+}
+
+
+static BSON_INLINE void *
+bson_atomic_ptr_fetch (void *volatile const *ptr, enum bson_atomic_memorder ord)
```

Comment:
Agcchgh. I'm not sure how I keep missing warnings, as I have `-Werror` turned on?

Either way, it is because fetching a value is a non-modifying operation. There was another error regarding discarded-`const` in another branch. In this case where `_fetch()` is implemented as `_fetch_add` of with addend zero, I need to do a cast to promise the compiler that the `fetch_add()` is a no-op.